### PR TITLE
Add support for exposing accessibility identifier as resource-id on Android

### PR DESCRIPTION
**Pull request by bartekpacia**
Accompanying framework PR: https://github.com/flutter/flutter/pull/138331

This PR implements support for exposing `SemanticsProperties.identifier` on Android as `resource-id`. Mainly targeted at https://github.com/flutter/flutter/issues/17988. Would also fix https://github.com/flutter/flutter/issues/issues/137735 but it was marked as duplicate. Anyway, there's a lot of context in that issue.

This PR requires changing the `SemanticsUpdateBuilder` interface (defined in engine) that framework depends on, so it requires introducing a temporary API ([see question I asked on Discord](https://discord.com/channels/608014603317936148/608018585025118217/1174845658033819729) to learn more about this approach).

Steps:
**part 1: [engine] add `SemanticsUpdateBuilderNew`** <-- we are here
part 2: [flutter] use `SemanticsUpdateBuilderNew`
part 3: [engine] update `SemanticsUpdateBuilder` to be the same as `SemanticsUpdateBuilderNew`*
part 4: [flutter] use (now updated) `SemanticsUpdateBuilder` again.
part 5: [engine] remove `SemanticsBuilderNew`

I'd like to do these changes first, and only then continue with [the proper framework PR](https://github.com/flutter/flutter/pull/138331).

*More specifically: update `SemanticsUpdateBuilder.updateNode()` to be the same as `SemanticsUpdateBuilderNew.updateNode()`. Number of arguments that function takes is the only change.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or the PR is [test-exempt]. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by bartekpacia** [comment]
### Demo of a basic example

Used framework commit: https://github.com/flutter/flutter/pull/138331/commits/cdfd94cd8ca429554ba463bc30a50112c3480503

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        body: Center(
          child: Semantics(
            identifier: 'some-blue-container',
            label: 'A blue container',
            child: Container(
              width: 100,
              height: 100,
              color: Colors.blue,
            ),
          ),
        ),
      ),
    );
  }
}
```

Interesting part of `uiautomator dump` result:

```xml
<node index="0" text="" resource-id="some-blue-container" class="android.view.View" package="com.example.example" content-desc="A blue container" checkable="false" checked="false" clickable="false" enabled="true" focusable="true" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[409,1006][671,1268]" />
```

<details><summary>Full uiautomator dump result</summary>

```xml
<?xml version='1.0' encoding='UTF-8' standalone='yes' ?>
<hierarchy rotation="0">
    <node index="0" text="" resource-id="" class="android.widget.FrameLayout" package="com.example.example" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,0][1080,2274]">
        <node index="0" text="" resource-id="" class="android.widget.LinearLayout" package="com.example.example" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,0][1080,2274]">
            <node index="0" text="" resource-id="android:id/content" class="android.widget.FrameLayout" package="com.example.example" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,0][1080,2274]">
                <node index="0" text="" resource-id="" class="android.widget.FrameLayout" package="com.example.example" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="true" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,0][1080,2274]">
                    <node index="0" text="" resource-id="" class="android.view.View" package="com.example.example" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,0][1080,2274]">
                        <node index="0" text="" resource-id="" class="android.view.View" package="com.example.example" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,0][1080,2274]">
                            <node index="0" text="" resource-id="" class="android.view.View" package="com.example.example" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,0][1080,2274]">
                                <node index="0" text="" resource-id="" class="android.view.View" package="com.example.example" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,0][1080,2274]">
                                    <node index="0" text="" resource-id="some-blue-container" class="android.view.View" package="com.example.example" content-desc="A blue container" checkable="false" checked="false" clickable="false" enabled="true" focusable="true" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[409,1006][671,1268]" />
                                </node>
                            </node>
                        </node>
                    </node>
                </node>
            </node>
        </node>
        <node index="1" text="" resource-id="android:id/statusBarBackground" class="android.view.View" package="com.example.example" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,0][1080,63]" />
        <node index="2" text="" resource-id="android:id/navigationBarBackground" class="android.view.View" package="com.example.example" content-desc="" checkable="false" checked="false" clickable="false" enabled="true" focusable="false" focused="false" scrollable="false" long-clickable="false" password="false" selected="false" bounds="[0,0][0,0]" />
    </node>
</hierarchy>
```

</details>

File: lib/ui/semantics.dart [hunk 1]
```diff
@@ -730,6 +730,9 @@ abstract class SemanticsUpdateBuilder {
   /// [PlatformDispatcher.onSemanticsActionEvent] callback might be called with
   /// an action that is no longer possible.
   ///
+  /// The `identifier` is a string that describes the node for UI automation.
```
Comment (chunhtai) [inline lib/ui/semantics.dart:733]: UI automation -> UI automation tools

maybe mention several example tools

File: lib/ui/semantics.dart [hunk 2]
```diff
@@ -872,6 +876,7 @@ base class _NativeSemanticsUpdateBuilder extends NativeFieldWrapperClass1 implem
     required double elevation,
     required double thickness,
     required Rect rect,
+    String? identifier,
```
Comment (chunhtai) [inline lib/ui/semantics.dart:879]: why nullable? can this be required?

File: lib/ui/semantics/semantics_update_builder.cc [hunk 3]
```diff
@@ -6,6 +6,7 @@
 
 #include <utility>
 
+#include "flutter/fml/logging.h"
```
Comment (chunhtai) [inline lib/ui/semantics/semantics_update_builder.cc:9]: debug code

File: lib/ui/semantics/semantics_update_builder.cc [hunk 4]
```diff
@@ -88,7 +90,12 @@ void SemanticsUpdateBuilder::updateNode(
                                SafeNarrow(right), SafeNarrow(bottom));
   node.elevation = elevation;
   node.thickness = thickness;
+  node.identifier = std::move(identifier);
   node.label = std::move(label);
+
+  // FML_LOG(ERROR) << "label: " << node.label
```
Comment (chunhtai) [inline lib/ui/semantics/semantics_update_builder.cc:96]: debug code

**Review by chunhtai — COMMENTED** [review]
Do you have iOS pr?

**Comment by bartekpacia** [comment]
Thank you for the review @chunhtai! This is not quite ready but I appreciate it. Also I'll ping you once it is ready.

> Do you have iOS pr?

Not yet.

File: lib/ui/semantics.dart (hunk 2 above)
Comment (bartekpacia) [reply lib/ui/semantics.dart:879]: tbh I'm not sure. I took inspiration from `tooltip` which is also a raw string, not `List<StringAttribute>`, and it is nullable.

But in the native semantics layer (at least on Android) the node's `resource-id` cannot be null, only empty.

File: lib/ui/semantics/semantics_update_builder.cc (hunk 3 above)
Comment (bartekpacia) [reply lib/ui/semantics/semantics_update_builder.cc:9]: fixed

File: lib/ui/semantics/semantics_update_builder.cc (hunk 4 above)
Comment (bartekpacia) [reply lib/ui/semantics/semantics_update_builder.cc:96]: fixed

**Comment by bartekpacia** [comment]
@chunhtai 

> Do you have iOS pr?

Should I implement the iOS part in the same PR as Android part, or can they happen in different ones?

**Comment by chunhtai** [comment]
I would suggest open a separate pr for iOS part and people can review entire thing as a whole.

**Comment by chunhtai** [comment]
Also is this apply to desktop?

**Comment by bartekpacia** [comment]
> Also can this apply to desktop?

On macOS – probably, it's pretty similar to mobile iOS. I don't know about other platforms though. But right now I'd like to bring this feature to mobile.

**Comment by bartekpacia** [comment]
This PR requires changing the `SemanticsUpdateBuilder` interface (defined in engine) that framework depends on. So it will require introducing a temporary API ([see question I asked on Discord](https://discord.com/channels/608014603317936148/608018585025118217/1174845658033819729)).

File: lib/ui/semantics.dart (hunk 1 above)
Comment (bartekpacia) [reply lib/ui/semantics.dart:733]: Yup, will do!

Comment (chunhtai) [reply lib/ui/semantics.dart:733]: Seems not addressed yet?

**Review by chunhtai — APPROVED** [review]
Just some nit picking. Otherwise LGTM, please only merge when all the relevant PRs are approved

File: lib/ui/semantics.dart (hunk 2 above)
Comment (bartekpacia) [reply lib/ui/semantics.dart]: Done in c50bceec24433b36860ba82a1388f35fbdd0dbff.

Should I do the same for `tooltip` (probably in a separate PR)?

File: lib/ui/semantics.dart (hunk 1 above)
Comment (bartekpacia) [reply lib/ui/semantics.dart]: yeah, I forgot haha

done in f760cfa5553724d6240242f9a213cd442d657a39

**Comment by bartekpacia** [comment]
> please only merge when all the relevant PRs are approved

Could you clarify this? This PR is actually the first one that has to land. I described the plan in the PR description ([link](https://github.com/flutter/engine/pull/47961#issue-1989602213)). (Also this is my first time doing a gradual migration like this, I'd appreciate you looking if that plan makes sense).

**Comment by chunhtai** [comment]
I meant you should have all prs approved, usually the framework and engine pr. We don't want to merge the engine pr and figure out the the code need to be updated due to addressing comment in the framework pr. It would also help to reduce the time that the transitional code lives in the code base

Would be good to have clean up pr up as well, but that is required if it is hard to do.

**Pull request merged**